### PR TITLE
JS: Add model of d3 selections

### PR DESCRIPTION
**Pull request by asgerf**
Adds a model of `d3` selections, which is the only part of `d3` where I could find anything relevant for our security queries.

`d3` selections are a bit like `jQuery` objects, in that they contain a collection of DOM elements, and have methods for interacting with them. The model just contributes news ways of interacting with the DOM via `d3` selections.

Evaluations: (internal links)
- [Evaluation d3 slugs](https://github.com/dsp-testing/asgerf-dca/tree/run/js/d36/reports) shows OK performance and a number of new XSS sinks and a few OK results.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/D3.qll [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+private import semmle.javascript.security.dataflow.Xss
+
+/** Provides classes and predicates modelling aspects of the `d3` library. */
+module D3 {
+  /** Gets an API node referring to the `d3` module. */
+  API::Node d3() {
+    result = API::moduleImport("d3")
+    or
+    result = API::moduleImport("d3-node").getInstance().getMember("d3")
+  }
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/D3.qll:13]: I had expected that `d3` would also be used as a global variable.

File: javascript/ql/src/semmle/javascript/security/dataflow/XssThroughDomCustomizations.qll [hunk 2]
```diff
@@ -51,14 +58,32 @@ module XssThroughDom {
...
+
   /**
    * A source for text from the DOM from a DOM property read or call to `getAttribute()`.
    */
   class DOMTextSource extends Source {
     DOMTextSource() {
       exists(DataFlow::PropRead read | read = this |
         read.getBase().getALocalSource() = DOM::domValueRef() and
-        read.mayHavePropertyName(["innerText", "textContent", "value", "name"])
+        read.mayHavePropertyName(unsafeDomPropertyName())
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/security/dataflow/XssThroughDomCustomizations.qll:84]: Unrelated to D3, but something I just thought of: 
The [`$().prop(propName)`](https://api.jquery.com/prop/) method might also be worth including here.

**Review by erik-krogh — COMMENTED** [review]
Nice. 

What is the status of recognizing the source of CVE-2017-1000006? 
I guess we don't recognize it.

File: javascript/ql/src/semmle/javascript/frameworks/D3.qll (hunk 1 above)
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/frameworks/D3.qll]: Good point, we should totally recognize that.

File: javascript/ql/src/semmle/javascript/security/dataflow/XssThroughDomCustomizations.qll (hunk 2 above)
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/security/dataflow/XssThroughDomCustomizations.qll:95]: Adding it in here and running another evaluation.

**Comment by asgerf** [comment]
> What is the status of recognizing the source of CVE-2017-1000006?

Oh thanks for brining that up. Added a few things that I missed from the model. We do find the sources now.

**Comment by asgerf** [comment]
New evaluations [on d3 slugs](https://github.com/dsp-testing/asgerf-dca/issues/16) and [on default slugs](https://github.com/dsp-testing/asgerf-dca/issues/17) look mostly OK, but I'll take a bit of time to investigate the slowest ones.

**Comment by asgerf** [comment]
As far as I can tell the outliers are just noise.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**